compiler/rustc_mir/src/transform/check_consts/validation.rs

```diff
@@ -636,17 +634,47 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 _,
             ) => self.check_op(ops::FnPtrCast),
 
-            Rvalue::Cast(CastKind::Pointer(PointerCast::Unsize), _, cast_ty) => {
-                if let Some(TypeAndMut { ty, .. }) = cast_ty.builtin_deref(true) {
-                    let unsized_ty = self.tcx.struct_tail_erasing_lifetimes(ty, self.param_env);
+            Rvalue::Cast(CastKind::Pointer(PointerCast::Unsize), _, _) => {
+                // With casts we are extra paranoid, and reject them when `dyn Trait` is involved
+                // even if the cast result is definitely unused. If it were used,
+                // `check_local_or_return_ty` would catch it. We cannot call
+                // `check_local_or_return_ty` here as taht would also reject things like unused
+                // mutable referencs that are allowed.
```

Comment:
I am not sure if we really want to be super paranoid here, but there is an explicit test for this behavior in `min_const_fn`:
```rust
const fn really_no_traits_i_mean_it() { (&() as &dyn std::fmt::Debug, ()).1 }
```
I think there is zero harm in accepting such code (given in particular that we accept it in `const`/`static` bodies). I'd prefer to just do no check at all here in the `Unsize` cast block; the existing checks will still error if any MIR local/temporary has a type involving `dyn Trait`. This would also reduce the amount of duplicate errors we emit.

Reply:
Yea, there's no need for these duplicate checks

Reply:
Ah turns out I was wrong and this function is actually still rejected (I just misinterpreted some compiletest output).